git/cmd.py

```diff
@@ -603,6 +610,73 @@ def execute(self, command,
         else:
             return stdout_value
 
+    def set_environment(self, **kwargs):
+        """
+        Set environment variables for future git invocations. Return all changed
+        values in a format that can be passed back into this function to revert
+        the changes:
+
+        ``Examples``::
+
+            old_env = self.set_environment(PWD='/tmp')
+            self.set_environment(**old_env)
+
+        :param kwargs: environment variables to use for git processes
+        :return: dict that maps environment variables to their old values
+        """
+        old_env = {}
+        for key, value in kwargs.iteritems():
+            # set value if it is None
+            if value is not None:
+                if key in self._environment:
+                    old_env[key] = self._environment[key]
+                else:
+                    old_env[key] = None
+                self._environment[key] = value
+            # remove key from environment if its value is None
+            elif key in self._environment:
+                old_env[key] = self._environment[key]
+                del self._environment[key]
+        return old_env
+
+    @contextmanager
+    def environment(self, **kwargs):
```

Comment:
This is an interesting one ! When used as `with repo.git.environment(FOO='bar'):`, the name makes total sense. When I imagine to write a test-case, I'd probably want to query the current state of the custom environment like this too: `repo.git.environment()`, following the `set_foo()`, `foo()` convention.

Can you imagine a better name for `environment()` to get a context manger ? I know, it's sweet and short already and anything else will make it longer ... `with altered_environment()`, `with changed_environment()`, `with custom_environment()` ?